src/loader/tsbgw_launcher.c

```diff
@@ -0,0 +1,511 @@
+
+						WaitForBackgroundWorkerStartup(tsbgw_he->db_scheduler_handle, &worker_pid);
+						ereport(LOG, (errmsg("Worker started with PID %d", worker_pid)));
+					}
+				}
+				tsbgw_message_send_ack(message, true);
+				break;
+			case STOP:
+				ereport(LOG, (errmsg("in stop condition")));
+
+				tsbgw_he = hash_search(db_htab, &message->db_oid, HASH_REMOVE, &found);
+				if (found)
+				{
+					TerminateBackgroundWorker(tsbgw_he->db_scheduler_handle);
+					WaitForBackgroundWorkerShutdown(tsbgw_he->db_scheduler_handle);
+					tsbgw_total_workers_decrement();
+				}
+				tsbgw_message_send_ack(message, true);
+				break;
+
+			case RESTART:
+				ereport(LOG, (errmsg("in restart condition")));
+
+				tsbgw_he = hash_search(db_htab, &message->db_oid, HASH_ENTER, &found);
+				if (found)
+				{
+					TerminateBackgroundWorker(tsbgw_he->db_scheduler_handle);
+					WaitForBackgroundWorkerShutdown(tsbgw_he->db_scheduler_handle);
+				}
+				else if (!tsbgw_total_workers_increment(TSBGW_MAX_WORKERS_GUC_STANDIN))
+					ereport(ERROR, (errmsg("timescale background worker could not be started")));
+
+				worker_registered = register_tsbgw_entrypoint_for_db(tsbgw_he->db_oid, vxid, &tsbgw_he->db_scheduler_handle);
+				if (worker_registered)
+				{
+					pid_t		worker_pid;
+
+					WaitForBackgroundWorkerStartup(tsbgw_he->db_scheduler_handle, &worker_pid);
+					ereport(LOG, (errmsg("Worker started with PID %d", worker_pid)));
+				}
+                tsbgw_message_send_ack(message, true);
+				break;
+			default:
+				ereport(LOG, ((errmsg("timescalebgw cluster launcher unexpected message received."))));
+		}
+	}
+}
+
+extern void
+tsbgw_cluster_launcher_main(void)
+{
+
+	int         n;
+	HTAB	    *db_htab;
+
+	pqsignal(SIGTERM, tsbgw_sigterm);
+	BackgroundWorkerUnblockSignals();
+	if (!tsbgw_total_workers_increment(TSBGW_MAX_WORKERS_GUC_STANDIN))
+	{
+		/*
+		 * should be the first thing happening so if we already exceeded our
+		 * limits it means we have a limit of 0 and we should just exit
+		 */
+		ereport(LOG, (errmsg("Timescale Background Worker Limit Set To %d. Please set higher if you would like to use Background Workers.",TSBGW_MAX_WORKERS_GUC_STANDIN)));
+		proc_exit(0);
+	}
+	/* Connect to the db, no db name yet, so can only access shared catalogs */
+	BackgroundWorkerInitializeConnection(NULL, NULL);
+    pgstat_report_appname("Timescale BGW Cluster Launcher");
+	db_htab = populate_database_htab();
+
+    for (n = 1; n <= hash_get_num_entries(db_htab); n++ )
+        if (!tsbgw_total_workers_increment(TSBGW_MAX_WORKERS_GUC_STANDIN))
+        {
+            ereport(LOG, (errmsg("Total databases = %ld Timescale Background Worker Limit Set To %d. Please set higher if you would like to use Background Workers.", hash_get_num_entries(db_htab), TSBGW_MAX_WORKERS_GUC_STANDIN)));
+            proc_exit(0);
+        }
+	start_db_schedulers(db_htab);
+
+
+	while (!got_sigterm)
+	{
+		int			wl_rc;
+
+		check_for_stopped_db_schedulers(db_htab);
```

Comment:
There are 4 ways that a scheduler can stop (that I can think of anyway)
1) it exits by itself for a good reason (basically timescale isn't installed in db) 
2) it crashes, but potentially corrupts shared memory/segfaults, so it causes all procs to restart and a reload of shared state.
3) it crashes but doesn't corrupt anything, so doesn't cause a restart. This may be impossible. 
4) It exits due to receipt of signal from the loader. 
The loader will be notified at worker exit because we've registered to be so notified, but we don't get told which one has exited so we have to go through and check who's still alive. However, on stop, we know we have removed it because we wait for the worker to shut down, so we can remove it immediately, but otherwise we need to go through and garbage collect.